tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java

```diff
@@ -0,0 +1,134 @@
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package io.ballerina.test;
+
+import org.ballerinalang.model.symbols.SymbolKind;
+import org.ballerinalang.test.BCompileUtil;
+import org.ballerinalang.test.CompileResult;
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import org.wso2.ballerinalang.compiler.semantics.analyzer.Types;
+import org.wso2.ballerinalang.compiler.semantics.model.Scope;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.BSymbol;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.BTypeDefinitionSymbol;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BType;
+import org.wso2.ballerinalang.compiler.tree.BLangPackage;
+import org.wso2.ballerinalang.compiler.util.CompilerContext;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static io.ballerina.test.utils.Constants.COMMENT;
+import static io.ballerina.test.utils.Constants.FAILING_FILE;
+import static io.ballerina.test.utils.Constants.SUBTYPE_SYMBOL;
+
+/**
+ * Test semantic type relations.
+ *
+ * @since 2.0.0
+ */
+public class SemTypeTest {
+
+    @DataProvider(name = "filePathProvider")
+    public Object[] filePathProvider() {
+        File dir = resolvePath("test-src").toFile();
+        List<String> files = new ArrayList<>();
+        for (File file : Objects.requireNonNull(dir.listFiles())) {
+            // This check is used to remove failing tests.
+            if (!file.getName().contains(FAILING_FILE)) {
+                files.add(file.getAbsolutePath());
+            }
+        }
+        return files.toArray();
+    }
+
+    @Test(dataProvider = "filePathProvider")
+    public void testSubTypeRelationship(String filePath) throws IOException {
+        List<String> actualRelations = actualSubTypeRelations(filePath);
+        List<String> expectedRelations = expectedSubTypeRelations(filePath);
+        Assert.assertEquals(actualRelations.size(), expectedRelations.size());
+        for (String actualRelation : actualRelations) {
+            expectedRelations.removeIf(expectedRelation -> Objects.equals(actualRelation, expectedRelation));
+        }
+        Assert.assertEquals(expectedRelations.size(), 0);
+    }
+
+    private List<String> actualSubTypeRelations(String filePath) {
+        CompileResult compileResult = BCompileUtil.compile(filePath);
+        BLangPackage bLangPackage = (BLangPackage) compileResult.getAST();
+        List<BTypeDefinitionSymbol> bTypeDefinitionSymbols = new ArrayList<>();
+        for (Scope.ScopeEntry value : bLangPackage.symbol.scope.entries.values()) {
+            BSymbol bSymbol = value.symbol;
+            if (bSymbol.kind == SymbolKind.TYPE_DEF) {
```

Comment:
@KavinduZoysa We need to skip the generated type definitions
check the following example
```ballerina
// ArrayFiveOfIntFive<:ArrayOfIntFive
// ROArrayFiveOfIntFive<:ArrayOfIntFive
// ROArrayFiveOfIntFive<:ArrayFiveOfIntFive

const FIVE = 5;

type ArrayOfIntFive int[][FIVE];

type ArrayFiveOfIntFive int[FIVE][FIVE];

type ROArrayFiveOfIntFive ArrayFiveOfIntFive & readonly;

type INT int;
```

Reply:
@chiranSachintha, what do you mean by  `generated type definitions`? 
In this example, we cannot find usages of `type INT int;`.

Reply:
for the `const FIVE = 5;` we generate type definition(with anonymous name) by the compiler. Here, don't we need to ignore those type definitions?

Reply:
Fixed with https://github.com/ballerina-platform/ballerina-lang/pull/34704